Documentation/anti-patterns.md

```diff
@@ -0,0 +1,195 @@
+  For example, if the error is an `INVALID_ARGUMENT` error, that probably means your
+  service has a bug (otherwise it shouldn't have triggered this error), in which
+  case `INTERNAL` is more appropriate to return back to your users.
+
+### Example: Handling errors from an RPC
+
+The following code snippet demonstrates how to handle errors from an RPC in
+gRPC:
+
+```go 
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
+
+res, err := client.MyRPC(ctx, &MyRequest{})
+if err != nil {
+    // Handle the error appropriately,
+    // log it & return an error to the caller, etc.
+    log.Printf("Error calling MyRPC: %v", err)
+    return nil, err
+}
+
+// Use the response as appropriate 
+log.Printf("MyRPC response: %v", res)
+
+```
+
+To determine the type of error that occurred, you can use the status field of
+the error response:
+
+
+```go 
+resp, err := client.MakeRPC(context.Background(), request) 
+if err != nil {
+  status, ok := status.FromError(err) 
+  if ok {
+    // Handle the error based on its status code 
+    if status.Code() == codes.NotFound {
+      log.Println("Requested resource not found")
+    } else {
+      log.Printf("RPC error: %v", status.Message())
+    }
+  } else {
+    //Handle non-RPC errors 
+    log.Printf("Non-RPC error: %v", err)
+  }
+  return
+}        
+
+// Use the response as needed 
+log.Printf("Response received: %v", resp) 
+```
+
+### Example: Using a backoff strategy
+
+
+When retrying failed RPCs, use a backoff strategy to avoid overwhelming the
+server or exacerbating network issues:
+
+
+```go 
+var res *MyResponse
+var err error
+
+// If the user doesn't have a context with a deadline, create one
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
+
+// Retry the RPC call a maximum number of times
+for i := 0; i < maxRetries; i++ {
+    
+    // Make the RPC call
+    res, err = client.MyRPC(ctx, &MyRequest{})
+    
+    // Check if the RPC call was successful
+    if err == nil {
+        // The RPC was successful, so break out of the loop
+        break
+    }
+    
+    // The RPC failed, so wait for a backoff period before retrying
+    backoff := time.Duration(i) * time.Second
+    log.Printf("Error calling MyRPC: %v; retrying in %v", err, backoff)
+    time.Sleep(backoff)
+}
+
+// Check if the RPC call was successful after all retries
+if err != nil {
+    // All retries failed, so handle the error appropriately
+    log.Printf("Error calling MyRPC: %v", err)
+    return nil, err
+}
+
+// Use the response as appropriate
+log.Printf("MyRPC response: %v", res)
+
+```
+
+
+## Conclusion
+
+The [`FailOnNonTempDialError`](https://pkg.go.dev/google.golang.org/grpc#FailOnNonTempDialError), [`WithBlock`](https://pkg.go.dev/google.golang.org/grpc#WithBlock), and [`WithReturnConnectionError`](https://pkg.go.dev/google.golang.org/grpc#WithReturnConnectionError)
```

Comment:
added to both the first references at the beginning and the references at the very end.
@dfawley

Reply:
Could you please add the references in the other direction?  It would be nice if the godoc for those functions pointed to this document.

Reply:
For better understanding, do you want me to add godoc references to all other docstrings?

Reply:
I would like a link to this document from https://github.com/grpc/grpc-go/blob/2997e84fd8d18ddb000ac6736129b48b3c9773ec/dialoptions.go#L297 (etc).  I recognize this document has no current link, so it will look broken in the PR, but it will work when it's merged.  (https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md)

Reply:
Got it, i'll prepare new PR for that

Reply:
You can do it in this one, thanks.